File: tradingview-python-connector/exchange/exchange.py
from abc import ABC, abstractmethod

from order.order import Order
from order.position import Position


class Exchange(ABC):
    """
    An abstract base class representing an exchange interface.
    """

    @abstractmethod
    def place_order(
        self,
        asset: str,
        order: Order,
        reduce_only: bool = False,
    ) -> dict:
        """
        Place an order on the exchange.

        Args:
            asset (str): The asset symbol to trade.
            reduce_only (dict): Whether the order should only reduce a position, not increase/reverse it.
            order (Order): Order object containing price and size info.

        Returns:
            dict: The result of the order placement.
        """
        pass

    @abstractmethod
    def cancel_existing_orders(self, asset: str) -> bool:
        """
        Cancel existing orders for the given asset.

        Args:
            asset (str): The asset symbol for which to cancel orders.

        Returns:
            bool: True if orders were successfully canceled, False otherwise.
        """
        pass

    @abstractmethod
    def close_positions(self, asset: str) -> bool:
        """
        Close existing positions for the given asset.

        Args:
            asset (str): The asset symbol for which to close positions.

        Returns:
            bool: True if positions were successfully closed, False otherwise.
        """
        pass

    @abstractmethod
    def has_open_position(self, asset: str) -> Position:
        """
        Check if there is an open position for the given asset.

        Args:
            asset (str): The asset symbol to check.

        Returns:
            bool: True if there is an open position, False otherwise.
        """
        pass

File: tradingview-rust-connector/src/exchanges/exchange.rs
use crate::types::Order;
use crate::types::Position;

pub trait Exchange {
    /// Place an order on the exchange.
    ///
    /// # Arguments
    ///
    /// * `asset` - The asset symbol to trade.
    /// * `reduce_only` - Whether the order should only reduce a position, not increase/reverse it.
    /// * `order` - Order object containing price and size info.
    ///
    /// # Returns
    ///
    /// A dictionary containing the result of the order placement.
    async fn place_order(&self, order: Order) -> anyhow::Result<serde_json::Value>;

    /// Cancel existing orders for the given asset.
    ///
    /// # Arguments
    ///
    /// * `asset` - The asset symbol for which to cancel orders.
    ///
    /// # Returns
    ///
    /// `true` if orders were successfully canceled, `false` otherwise.
    async fn cancel_existing_orders(&self, asset: &str) -> anyhow::Result<bool>;

    /// Close existing positions for the given asset.
    ///
    /// # Arguments
    ///
    /// * `asset` - The asset symbol for which to close positions.
    ///
    /// # Returns
    ///
    /// `true` if positions were successfully closed, `false` otherwise.
    fn close_positions(&self, asset: &str) -> bool;

    /// Check if there is an open position for the given asset.
    ///
    /// # Arguments
    ///
    /// * `asset` - The asset symbol to check.
    ///
    /// # Returns
    ///
    /// `true` if there is an open position, `false` otherwise.
    async fn has_open_position(&self, asset: &str) -> anyhow::Result<Position>;
}